gdx/src/com/badlogic/gdx/graphics/debugging/GLDebugger.java

```diff
@@ -0,0 +1,111 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ ******************************************************************************/
+
+package com.badlogic.gdx.graphics.debugging;
+
+import static com.badlogic.gdx.graphics.GL20.*;
+
+import com.badlogic.gdx.Gdx;
+import com.badlogic.gdx.utils.GdxRuntimeException;
+
+/** Enabling this will wrap Gdx.gl20 and Gdx.gl30 instances into GLDebugger subclasses that check for GL errors after each GL call,
+ * using glGetError. When error happens, {@link GLDebugger#listener} is called inside GL call, so the stack trace may be
+ * inspected.
+ *
+ * @see GL30Debugger
+ * @see GL20Debugger
+ * @see com.badlogic.gdx.graphics.profiling.GLProfiler
+ * @author Jan Polák */
+public abstract class GLDebugger {
+
+	/** Listener that will log using Gdx.app.error GL error number and GL function. */
+	public static final GLDebuggerErrorListener LOGGING_LISTENER = new GLDebuggerErrorListener() {
+		@Override
+		public void onError (int error) {
+			String place = null;
+			try {
+				StackTraceElement[] stack = Thread.currentThread().getStackTrace();
+				if (stack.length >= 4) {
+					final StackTraceElement glMethod = stack[3];
+					place = glMethod.getMethodName();
+				}
+			} catch (Exception ignored) {
+			}
+
+			if (place != null) {
+				Gdx.app.error("GLDebugger", "Error " + resolveErrorNumber(error) + " from " + place);
+			} else {
+				Gdx.app.error("GLDebugger", "Error " + resolveErrorNumber(error) + " at: ", new Exception()); // This will capture
+// current stack trace for logging, if possible
+			}
+		}
+	};
+
+	/** Listener that will throw a GdxRuntimeException with error number. */
+	public static final GLDebuggerErrorListener THROWING_LISTENER = new GLDebuggerErrorListener() {
+		@Override
+		public void onError (int error) {
+			throw new GdxRuntimeException("GLDebugger: Got gl error " + resolveErrorNumber(error));
+		}
+	};
+
+	public static String resolveErrorNumber (int error) {
+		switch (error) {
+		case GL_INVALID_VALUE:
+			return "GL_INVALID_VALUE";
+		case GL_INVALID_OPERATION:
+			return "GL_INVALID_OPERATION";
+		case GL_INVALID_FRAMEBUFFER_OPERATION:
+			return "GL_INVALID_FRAMEBUFFER_OPERATION";
+		case GL_INVALID_ENUM:
+			return "GL_INVALID_ENUM";
+		case GL_OUT_OF_MEMORY:
+			return "GL_OUT_OF_MEMORY";
+		default:
+			return "number " + error;
+		}
+	}
+
+	/** This listener will be called when GLDebugger is enabled and any GL call sets error number (retrievable by glGetError call).
+	 *
+	 * Default is {@link GLDebugger#LOGGING_LISTENER}. */
+	public static GLDebuggerErrorListener listener = LOGGING_LISTENER;
+
+	/** Enables error checking by replacing the {@code GL20} and {@code GL30} instances with debugging ones. */
+	public static void enable () {
+		if (!isEnabled()) {
+			Gdx.gl30 = Gdx.gl30 == null ? null : new GL30Debugger(Gdx.gl30);
+			Gdx.gl20 = Gdx.gl30 != null ? Gdx.gl30 : new GL20Debugger(Gdx.gl20);

```

Comment:
enable() and disable() were carefully copied from GLProfiler, because it does basically the same thing and that code is already proven to work. There is one concern however, line [100 in file GLDebugger](https://github.com/libgdx/libgdx/pull/2889/files#diff-07cfe7d207b530cf767547cbbd911849R100) will probably fail when GL30 is enabled. That is because when GL30 is enabled, Gdx.gl30,gl20 and gl will all point to GL30Debugger. That class is not a subclass of GL20Debugger that is checked against here, so it will stay wrapped. This exact problem is present in GLProfiler.

Now that I'm looking at it even more carefully, it will fail even on the line before, when gl30 is enabled. Enabling is correct though. So unless I am missing something and the code is indeed correct, I'll patch it for both profiler and debugger. Should I do it in this PR or make a new one?